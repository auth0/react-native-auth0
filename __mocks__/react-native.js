/**
 * This file provides a project-wide, self-contained mock for the 'react-native' package.
 */

const React = require('react');

// A fake native module to be returned by TurboModuleRegistry.getEnforcing.
const mockNativeModule = {
  // ... (all the mock native methods from before)
  initializeAuth0WithConfiguration: jest.fn(() => Promise.resolve()),
  webAuth: jest.fn(() => Promise.resolve({})),
  webAuthLogout: jest.fn(() => Promise.resolve()),
  cancelWebAuth: jest.fn(() => Promise.resolve()),
  getBundleIdentifier: jest.fn(() => Promise.resolve('com.my.app')),
  resumeWebAuth: jest.fn(() => Promise.resolve(true)),
  saveCredentials: jest.fn(() => Promise.resolve()),
  getCredentials: jest.fn(() => Promise.resolve({})),
  hasValidCredentials: jest.fn(() => Promise.resolve(true)),
  clearCredentials: jest.fn(() => Promise.resolve()),
  hasValidInstance: jest.fn(() => Promise.resolve(true)),
};

// Export the mocked module.
module.exports = {
  // Mock the Platform module
  Platform: {
    OS: 'ios',
  },
  // Mock the TurboModuleRegistry
  TurboModuleRegistry: {
    getEnforcing: () => mockNativeModule,
  },
  // Mock the Linking module
  Linking: {
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
    getInitialURL: jest.fn(() => Promise.resolve(null)),
  },

  // FIX: Add mocks for core UI components used in tests.
  // We can use a simple functional component that just renders its children
  // and passes along any props.
  View: (props) => React.createElement('View', props, props.children),
  Text: (props) => React.createElement('Text', props, props.children),
  Button: (props) => React.createElement('Button', props, props.children),
  // Add any other components your tests might use, e.g., StyleSheet
  StyleSheet: {
    create: (styles) => styles,
  },
};
