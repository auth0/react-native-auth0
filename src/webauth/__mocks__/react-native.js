import Linking from './linking';
import A0Auth0 from './auth0';

const mock = {};

mock.Linking = new Linking();
mock.NativeModules = { A0Auth0: new A0Auth0() };
mock.Platform = { OS: 'test-os' };

module.exports = mock;
