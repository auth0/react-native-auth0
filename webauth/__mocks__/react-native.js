import Linking from './linking';
import A0Auth0 from './auth0';

const mock = {};

mock.Linking = new Linking();
mock.NativeModules = { A0Auth0: new A0Auth0() };
mock.Platform = { OS: "ios", Version: 11.1 };

module.exports = mock;