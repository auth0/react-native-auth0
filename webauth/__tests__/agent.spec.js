jest.mock('react-native');
import Agent from '../agent';
import { NativeModules, Linking } from 'react-native';
const A0Auth0 = NativeModules.A0Auth0;

describe('Agent', () => {
  const agent = new Agent();

  const callbackScheme = 'callback';
  const url = 'https://auth0.com';
  const authorizeUrl = 'https://auth0.com/authorize';

  describe('show', () => {

    describe('allowAuthenticationSession = false', () => {
    
      beforeEach(() => {
        NativeModules.A0Auth0 = A0Auth0;
        A0Auth0.reset();
      });

      it('should fail if native module is not linked', async () => {
        NativeModules.A0Auth0 = undefined;
        expect.assertions(1);
        await expect(agent.show({ url, callbackScheme })).rejects.toMatchSnapshot();
      });

      describe('complete web flow', () => {
        beforeEach(() => {
          A0Auth0.onUrl = () => {
            Linking.emitter.emit('url', { url });
          };
        });

        it('should call the correct method', async () => {
          await agent.show({ url, callbackScheme })
          expect(A0Auth0.showUrlCalled).toBeTruthy();
          expect(A0Auth0.showAuthorizationCalled).toBeFalsy();
        });

        it('should resolve promise with url result', async () => {
          expect.assertions(1);
          await expect(agent.show({ url, callbackScheme })).resolves.toMatchSnapshot();
        });

        it('should show initial url', async () => {
          expect.assertions(1);
          await agent.show({ url: authorizeUrl, callbackScheme });
          expect(A0Auth0.url).toEqual(authorizeUrl);
        });
      });

      describe('listeners', () => {

        it('should register url listeners', () => {
          A0Auth0.onUrl = () => {};
          agent.show({ url: authorizeUrl, callbackScheme });
          expect(Linking.emitter.listenerCount('url')).toEqual(1);
        });

        it('should remove url listeners when done', async () => {
          A0Auth0.onUrl = () => {
            Linking.emitter.emit('url', { url });
          };
          expect.assertions(1);
          await agent.show({ url: authorizeUrl, callbackScheme });
          expect(Linking.emitter.listenerCount('url')).toEqual(0);
        });

        it('should remove url listeners when load fails', async () => {
          expect.assertions(1);
          A0Auth0.error = new Error('failed to load');
          await agent.show({ url: authorizeUrl, callbackScheme }).catch((err) => Promise.resolve(err));
          expect(Linking.emitter.listenerCount('url')).toEqual(0);
        });

        it('should remove url listeners on first load', async () => {
          expect.assertions(1);
          await agent.show({ url: authorizeUrl, callbackScheme, closeOnLoad: true });
          expect(Linking.emitter.listenerCount('url')).toEqual(0);
        });

      });

      describe('failed flow', () => {

        it('should reject with error', async () => {
          expect.assertions(1);
          A0Auth0.error = new Error('failed to load');
          await expect(agent.show({ url, callbackScheme })).rejects.toMatchSnapshot();
        });

      });
    });

    describe('allowAuthenticationSession = true', () => {

      beforeEach(() => {
        NativeModules.A0Auth0 = A0Auth0;
        A0Auth0.reset();
      });

      it('should fail if native module is not linked', async () => {
        NativeModules.A0Auth0 = undefined;
        expect.assertions(1);
        await expect(agent.show({
          url,
          callbackScheme,
          allowAuthenticationSession: true
        })).rejects.toMatchSnapshot();
      });

      describe('complete authorization flow', () => {
        beforeEach(() => {
          A0Auth0.callbackUrl = url;
        });

        it('should call the correct method', async () => {
          await agent.show({
            url,
            callbackScheme,
            allowAuthenticationSession: true
          })
          expect(A0Auth0.showAuthorizationCalled).toBeTruthy();
          expect(A0Auth0.showUrlCalled).toBeFalsy();
        });

        it('should resolve promise with url result', async () => {
          expect.assertions(1);
          await expect(agent.show({
            url: authorizeUrl,
            callbackScheme,
            allowAuthenticationSession: true
          })).resolves.toMatchSnapshot();
        });

        it('should show initial url', async () => {
          expect.assertions(1);
          await agent.show({
            url: authorizeUrl,
            callbackScheme,
            allowAuthenticationSession: true
          });
          expect(A0Auth0.url).toEqual(authorizeUrl);
        });
      });

      describe('failed flow', () => {

        it('should reject with error', async () => {
          expect.assertions(1);
          A0Auth0.error = new Error('failed to load');
          await expect(agent.show({
            url,
            callbackScheme,
            allowAuthenticationSession: true
          })).rejects.toMatchSnapshot();
        });

      });
    });
  });

  describe('newTransaction', () => {
    
    it('should call native integration', async () => {
      const parameters = {state: 'state'};
      A0Auth0.parameters = parameters;
      expect.assertions(1);
      await expect(agent.newTransaction()).resolves.toMatchSnapshot();
    });

    it('should fail if native module is not linked', async () => {
      NativeModules.A0Auth0 = undefined;
      expect.assertions(1);
      await expect(agent.newTransaction()).rejects.toMatchSnapshot();
    });

  });
});
