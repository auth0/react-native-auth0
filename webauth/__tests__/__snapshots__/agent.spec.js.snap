// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Agent newTransaction should call native integration 1`] = `
Object {
  "state": "state",
}
`;

exports[`Agent newTransaction should fail if native module is not linked 1`] = `[Error: Missing NativeModule. Please make sure you run \`react-native link react-native-auth0\`]`;

exports[`Agent show allowAuthenticationSession = false complete web flow should resolve promise with url result 1`] = `"https://auth0.com"`;

exports[`Agent show allowAuthenticationSession = false failed flow should reject with error 1`] = `[Error: failed to load]`;

exports[`Agent show allowAuthenticationSession = false should fail if native module is not linked 1`] = `[Error: Missing NativeModule. Please make sure you run \`react-native link react-native-auth0\`]`;

exports[`Agent show allowAuthenticationSession = true complete authorization flow should resolve promise with url result 1`] = `"https://auth0.com"`;

exports[`Agent show allowAuthenticationSession = true failed flow should reject with error 1`] = `[Error: failed to load]`;

exports[`Agent show allowAuthenticationSession = true should fail if native module is not linked 1`] = `[Error: Missing NativeModule. Please make sure you run \`react-native link react-native-auth0\`]`;
